atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java

```diff
@@ -110,60 +114,79 @@ public void run() {
             Thread.sleep(getBackoffTimeWhenSweepHasNotRun());
             log.info("Starting background sweeper.");
             while (true) {
-                long millisToSleep = checkConfigAndRunSweep(locks);
-                Thread.sleep(millisToSleep);
+                SweepOutcome outcome = checkConfigAndRunSweep(locks);
+
+                updateBatchSize(outcome);
+                updateMetrics(outcome);
+
+                sleepUntilNextRun(outcome);
             }
         } catch (InterruptedException e) {
             log.warn("Shutting down background sweeper. Please restart the service to rerun background sweep.");
         }
     }
 
-    // Returns milliseconds to sleep
+    private void updateBatchSize(SweepOutcome outcome) {
+        if (outcome == SweepOutcome.SUCCESS) {
+            batchSizeMultiplier = Math.min(1.0, batchSizeMultiplier * 1.01);
+            return;
+        }
+        if (outcome == SweepOutcome.RETRYING_WITH_SMALLER_BATCH) {
+            SweepBatchConfig lastBatchConfig = specificTableSweeper.getAdjustedBatchConfig();
+
+            // Cut batch size in half, always sweep at least one row (we round down).
+            batchSizeMultiplier = Math.max(batchSizeMultiplier / 2, 1.5 / lastBatchConfig.candidateBatchSize());
+
+            log.warn("The background sweep job failed unexpectedly with candidate batch size {},"
+                            + " delete batch size {},"
+                            + " and {} cell+timestamp pairs to examine."
+                            + " Attempting to continue with new batchSizeMultiplier {}",
+                    SafeArg.of("candidateBatchSize", lastBatchConfig.candidateBatchSize()),
+                    SafeArg.of("deleteBatchSize", lastBatchConfig.deleteBatchSize()),
+                    SafeArg.of("maxCellTsPairsToExamine", lastBatchConfig.maxCellTsPairsToExamine()),
+                    SafeArg.of("batchSizeMultiplier", batchSizeMultiplier));
```

Comment:
We already adjusted the batch config, so sweep did not fail with this one as we will run sweep with this config only in the next iteration. The config used in the previous iteration will be `specificTableSweeper.getSweepBatchConfig()`.

Also, logging the batch size multiplier is not great info, lets log the `newBatchConfig` here.

Reply:
We haven't changed the batch config here.  We've changed the multiplier but we haven't recalculated the config.  We get the new config if we call `getAdjustedBatchConfig` after changing the multiplier which doesn't happen until the next iteration runs.

This is why mutability sucks, makes it harder to reason about this kind of thing.